group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java

```diff
@@ -1243,4 +1406,1283 @@ public static String consumerGroupSessionTimeoutKey(String groupId, String membe
+        GenericGroup group,
+        GenericGroupMember member,
+        long timeoutMs
+    ) {
+        String heartbeatKey = heartbeatKey(group.groupId(), member.memberId());
+
+        // Reschedule the next heartbeat expiration deadline
+        timer.schedule(heartbeatKey,
+            timeoutMs,
+            TimeUnit.MILLISECONDS,
+            false,
+            () -> expireGenericGroupMemberHeartbeat(group, member.memberId()));
+    }
+
+    /**
+     * Remove the sync key from the timer and clear all pending sync members from the group.
+     * Invoked when a new rebalance is triggered.
+     *
+     * @param group  The group.
+     */
+    private void removeSyncExpiration(GenericGroup group) {
+        group.clearPendingSyncMembers();
+        timer.cancel(syncKey(group.groupId()));
+    }
+
+    /**
+     * Expire pending sync.
+     *
+     * @param group           The group.
+     * @param generationId    The generation when the pending sync was originally scheduled.
+     *
+     * @return The coordinator result that will be appended to the log.
+     * */
+    private CoordinatorResult<Void, Record> expirePendingSync(
+        GenericGroup group,
+        int generationId
+    ) {
+        if (generationId != group.generationId()) {
+            log.error("Received unexpected notification of sync expiration for {} with an old " +
+                "generation {} while the group has {}.", group.groupId(), generationId, group.generationId());
+        } else {
+            if (group.isInState(DEAD) || group.isInState(EMPTY) || group.isInState(PREPARING_REBALANCE)) {
+                log.error("Received unexpected notification of sync expiration after group {} already " +
+                    "transitioned to {} state.", group.groupId(), group.stateAsString());
+            } else if (group.isInState(COMPLETING_REBALANCE) || group.isInState(STABLE)) {
+                if (!group.hasReceivedSyncFromAllMembers()) {
+                    Set<String> pendingSyncMembers = group.allPendingSyncMembers();
+                    pendingSyncMembers.forEach(memberId -> {
+                        group.remove(memberId);
+                        timer.cancel(heartbeatKey(group.groupId(), memberId));
+                    });
+
+                    log.debug("Group {} removed members who haven't sent their sync requests: {}",
+                        group.groupId(), pendingSyncMembers);
+
+                    return prepareRebalance(group, "Removing " + pendingSyncMembers + " on pending sync request expiration");
+                }
+            }
+        }
+
+        return EMPTY_RESULT;
+    }
+
+    /**
+     * Checks whether the group can accept a joining member.
+     *
+     * @param group      The group.
+     * @param memberId   The member.
+     *
+     * @return whether the group can accept a joining member.
+     */
+    private boolean acceptJoiningMember(GenericGroup group, String memberId) {
```

Comment:
Note: We will have to also verify the number of member after the group is loaded, I think. This is something for another PR but to keep in mind.

Reply:
to confirm, you're saying we should call `acceptJoiningMember` while loading members?

Reply:
No. I think that the current coordinator triggers a rebalance if the number of members is higher than the max when a group is loaded.

Reply:
updated and added a test case